File: tasks_011-020/task_011/js/solution_1.js
const fs = require('fs')
const path = require('path')

const table = fs.readFileSync(path.join(__dirname, '..', 'grid.txt'), 'utf8')
	.trim()
	.split('\n')
	.map(row => row.split(' '))
const numsCount = 4

let result = 0

for (let row = 0; row < table.length; row++) {
	for (let column = 0; column < table[row].length; column++) {

		// Back diag \
		if (table[row + numsCount - 1] !== undefined) {
			if (table[row + numsCount - 1][column + numsCount - 1] !== undefined) {
				let product = table[row][column]
				for (let num = 1; num < numsCount; num++)
					product *= table[row + num][column + num]
				if (product > result) result = product
			}
		}

		// Forth diag /
		if (table[row + numsCount - 1] !== undefined) {
			if (table[row + numsCount - 1][column - numsCount + 1] !== undefined) {
				let product = table[row][column]
				for (let num = 1; num < numsCount; num++)
					product *= table[row + num][column - num]
				if (product > result) result = product
			}
		}

		// Top down |
		if (table[row + numsCount - 1] !== undefined) {
			let product = table[row][column]
			for (let num = 1; num < numsCount; num++)
				product *= table[row + num][column]
			if (product > result) result = product
		}

		// Left right —
		if (table[row][column + numsCount - 1] !== undefined) {
			let product = table[row][column]
			for (let num = 1; num < numsCount; num++)
				product *= table[row][column + num]
			if (product > result) result = product
		}
	}
}

console.log('Result:', result) // OUTPUT: 70600674

File: tasks_011-020/task_011/go/solution_1.go
package solution

func FirstSolution() int {
	var result int

	for ri, row := range grid {
		for ci, columnValue := range row {
			// Left to right: -
			if ci+length-1 < len(row) {
				product := columnValue

				for i := 1; i < length; i++ {
					product *= grid[ri][ci+i]
				}

				if product > result {
					result = product
				}
			}

			// Top to bottom: |
			if ri+length-1 < len(grid) {
				product := columnValue

				for i := 1; i < length; i++ {
					product *= grid[ri+i][ci]
				}

				if product > result {
					result = product
				}
			}

			// Forth diagonal: /
			if ri+length-1 < len(grid) && ci-length-1 >= 0 {
				product := columnValue

				for i := 1; i < length; i++ {
					product *= grid[ri+i][ci-i]
				}

				if product > result {
					result = product
				}
			}

			// Back diagonal: \
			if ri+length-1 < len(grid) && ci+length-1 < len(row) {
				product := columnValue

				for i := 1; i < length; i++ {
					product *= grid[ri+i][ci+i]
				}

				if product > result {
					result = product
				}
			}
		}
	}

	return result
}
